docs/book/src/providers/implementers-guide/controllers_and_reconciliation.md

```diff
@@ -0,0 +1,254 @@
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+```
+
+By returning an error, we request that our controller will get `Reconcile()` called again.
+That may not always be what we want - what if the object's been deleted? So let's check that:
+
+```
+var cluster infrastructurev1alpha3.MailgunCluster
+if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+    // 	import apierrors "k8s.io/apimachinery/pkg/api/errors"
+    if apierrors.IsNotFound(err) {
+        return ctrl.Result{}, nil
+    }
+    return ctrl.Result{}, err
+}
+```
+
+Now, if this were any old `kubebuilder` project we'd be done, but in our case we have one more object to retrieve.
+Cluster API splits a cluster into two objects: the [`Cluster` defined by Cluster API itself][cluster]. 
+We'll want to retrieve that as well. 
+Luckily, cluster API [provides a helper for us][getowner].
+
+```go
+cluster, err := util.GetOwnerMachine(ctx, r.Client, &mg)
+if err != nil {
+    return ctrl.Result{}, err
+
+}
+```
+
+### client-go versions
+At the time this document was written, `kubebuilder` pulls `client-go` version `1.14.1` into `go.mod` (it looks like `k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible`).
+
+If you encounter an error when compiling like:
+
+```
+../pkg/mod/k8s.io/client-go@v11.0.1-0.20190409021438-1a26190bd76a+incompatible/rest/request.go:598:31: not enough arguments in call to watch.NewStreamWatcher
+    have (*versioned.Decoder)
+    want (watch.Decoder, watch.Reporter)`
+```
+
+You may need to bump `client-go`. At time of writing, that means `1.15`, which looks like: `k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible`.
+## The fun part
+
+_More Documentation: [The Kubebuilder Book][book] has some excellent documentation on many things, including [how to write good controllers!][implement]_
+
+[book]: https://book.kubebuilder.io/
+[implement]: https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html
+
+Now that we have our objects, it's time to do something with them! 
+This is where your provider really comes into it's own.
+In our case, let's try sending some mail:
+
+```go
+subject := fmt.Sprintf("[%s] New Cluster %s requested", mgCluster.Spec.Priority, cluster.Name)
+body := fmt.Sprint("Hello! One cluster please.\n\n%s\n", mgCluster.Spec.Request)
+
+msg := mailgun.NewMessage(mgCluster.Spec.Requester, subject, body, r.Recipient)
+_, _, err = r.Mailgun.Send(msg)
+if err != nil {
+    return ctrl.Result{}, err
+}
+```
+
+## Idempotency
+
+But wait, this isn't quite right. 
+`Reconcile()` gets called periodically for updates, and any time any updates are made.
+That would mean we're potentially sending an email every few minutes!
+This is an important thing about controllers: they need to be [*idempotent*][idempotent].
+
+So in our case, we'll store the result of sending a message, and then check to see if we've sent one before.
+
+```
+if mgCluster.Status.MessageID != nil {
```

Comment:
I'd recommend putting the `MessageID` in `Spec` and not `Status`. Anything that we can't recreate, either with some sort of deterministic generation method or by querying/observing actual state, needs to be in `Spec`. This is to support proper disaster recovery of resources. If you have a backup of your `MailgunCluster` and you want to restore it, Kubernetes doesn't let you restore both spec & status together. Anything you have in status is discarded by the Kubernetes apiserver. It _is_ possible to restore your resource (spec only), then follow up with an update to restore the status, but that is two separate operations, and it's entirely possible (and likely) that your controller acts on the restored resource (spec only), possibly generating another email, creating another virtual machine, etc.